stumpy/mstump.py

```diff
@@ -161,8 +164,108 @@ def _get_multi_QT(start, T, m):
 
 
 @njit(parallel=True, fastmath=True)
+def _compute_multi_D(
+    d, k, idx, D, T, m, excl_zone, M_T, Σ_T, QT_even, QT_odd, QT_first, μ_Q, σ_Q
+):
+    for i in range(d):
+        # Numba's prange requires incrementing a range by 1 so replace
+        # `for j in range(k-1,0,-1)` with its incrementing compliment
+        for rev_j in prange(1, k):
+            j = k - rev_j
+            # GPU Stomp Parallel Implementation with Numba
+            # DOI: 10.1109/ICDM.2016.0085
+            # See Figure 5
+            if idx % 2 == 0:
+                # Even
+                QT_even[i, j] = (
+                    QT_odd[i, j - 1]
+                    - T[i, idx - 1] * T[i, j - 1]
+                    + T[i, idx + m - 1] * T[i, j + m - 1]
+                )
+            else:
+                # Odd
+                QT_odd[i, j] = (
+                    QT_even[i, j - 1]
+                    - T[i, idx - 1] * T[i, j - 1]
+                    + T[i, idx + m - 1] * T[i, j + m - 1]
+                )
+
+        if idx % 2 == 0:
+            QT_even[i, 0] = QT_first[i, idx]
+            D[i] = core._calculate_squared_distance_profile(
+                m, QT_even[i], μ_Q[i, idx], σ_Q[i, idx], M_T[i], Σ_T[i]
+            )
+        else:
+            QT_odd[i, 0] = QT_first[i, idx]
+            D[i] = core._calculate_squared_distance_profile(
+                m, QT_odd[i], μ_Q[i, idx], σ_Q[i, idx], M_T[i], Σ_T[i]
+            )
+
+    zone_start = max(0, idx - excl_zone)
+    zone_stop = min(k, idx + excl_zone)
+    D[:, zone_start : zone_stop + 1] = np.inf
+
+    # D = np.sqrt(D)
+
+
+_D = None
```

Comment:
Should this be de-indented here? It kinda looks like it's the middle of nowhere and is therefore a global variable?

Reply:
As far as I have understood multiprocessing with numpy, this has to be a global variable, which is why I made it "private".
The idea is to allocate shared memory using `multiprocessing.array` and then wrapping this chunk of memory with a global numpy array using `np.frombuffer`. This way all spawned processes get access to the same chunk of memory instead of everyone getting its own copy.